src/proposal.jl

```diff
@@ -26,29 +34,83 @@ function SliceNUTS(h::Hamiltonian, θ::AbstractVector{T}) where {T<:Real}
     return SliceNUTS(Leapfrog(find_good_eps(h, θ)))
 end
 
+function find_good_eps(rng::AbstractRNG, h::Hamiltonian, θ::AbstractVector{T}; max_n_iters::Int=100) where {T<:Real}
+    ϵ′ = ϵ = 0.1
+    a_min, a_cross, a_max = 0.25, 0.5, 0.75 # minimal, crossing, maximal accept ratio
+    d = 2.0
+
+    r = rand_momentum(rng, h)
+    H = hamiltonian_energy(h, θ, r)
+
+    θ′, r′, _is_valid = step(Leapfrog(ϵ), h, θ, r)
+    H_new = _is_valid ? hamiltonian_energy(h, θ′, r′) : Inf
+
+    ΔH = H - H_new
+    direction = ΔH > log(a_cross) ? 1 : -1
+
+    # Crossing step: increase/decrease ϵ until accept ratio cross a_cross.
+    for _ = 1:max_n_iters
+        ϵ′ = direction == 1 ? d * ϵ : 1 / d * ϵ
+        θ′, r′, _is_valid = step(Leapfrog(ϵ′), h, θ′, r′)
+        H_new = _is_valid ? hamiltonian_energy(h, θ′, r′) : Inf
+
+        ΔH = H - H_new
+        DEBUG && @debug "Crossing step" direction H_new ϵ "α = $(min(1, exp(ΔH)))"
+        if (direction == 1) && !(ΔH > log(a_cross))
+            break
+        elseif (direction == -1) && !(ΔH < log(a_cross))
+            break
+        else
+            ϵ = ϵ′
+        end
+    end
+
+    # Bisection step: ensure final accept ratio: a_min < a < a_max.
+    # See https://en.wikipedia.org/wiki/Bisection_method
+    ϵ, ϵ′ = ϵ < ϵ′ ? (ϵ, ϵ′) : (ϵ′, ϵ)  # ensure ϵ < ϵ′
+    for _ = 1:max_n_iters
+        ϵ_mid = middle(ϵ, ϵ′)
+        θ′, r′, _is_valid = step(Leapfrog(ϵ_mid), h, θ, r)
+        H_new = _is_valid ? hamiltonian_energy(h, θ′, r′) : Inf
+
+        ΔH = H - H_new
+        DEBUG && @debug "Bisection step" H_new ϵ_mid "α = $(min(1, exp(ΔH)))"
+        if (exp(ΔH) > a_max)
+            ϵ = ϵ_mid
+        elseif (exp(ΔH) < a_min)
+            ϵ′ = ϵ_mid
+        else
+            ϵ = ϵ_mid
+            break
+        end
+    end
+
+    return ϵ
+end
```

Comment:
This is copied from your updated version with interface adapted to this package's. Would be nice for you to double check if the porting is correct.

Reply:
sure